### PR TITLE
Add Turborepo and common npm scripts

### LINKED ISSUE #5680: Use Turborepo or other tooling for smart scripts running across packages
Improve script running using Turborepo or other related tooling. By leveraging smarter tooling, we should achieve the following for the monorepo:

- Incremental builds
- Content-aware hashing
- Remote Caching
- Parallel execution
- Zero runtime overhead
- Pruned subsets
- Task pipelines
- Profile in your browser (Turborepo only)
- VS Code Extension (wireit only)

## Tooling options

- [Turborepo](https://turborepo.org/)
- [wireit](https://github.com/google/wireit)
- [Chomp](https://chompbuild.com/)
- [Bazel](https://bazel.build/)

### PR DESCRIPTION
**Pull request by sam-b-rose**
### WHY are these changes introduced?

Fixes #5680 

### WHAT is this pull request doing?

Initial addition of [Turborepo](https://turborepo.org/) to make the following improvements:

- Running package scripts in parallel
- Caching build output
- Declarative control of task pipelines

This PR also updates the packages to use the local workspace version of the package for development (ie `*`). This informs the [topological dependency](https://turborepo.org/docs/core-concepts/pipelines#topological-dependency) of what to build first.

Also, this PR moves the `stylelint-polaris` content into a `src` folder for easier linting. We should also consider creating a src directory for all packages as a standard convention.

## Testing this PR

Try out these scripts. And try them twice to see the amazing awesomeness of caching!

```sh
yarn run build   # builds all packages
yarn run lint      # lints all packages
yarn run test     # runs tests for all packages
yarn run clean  # cleans all build output and node_modules
```

I'd like to also have a `dev` command that will run the website with the dependencies running in watch mode, but we need to create a `watch` build command for `polaris-react` first.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,12 +14,12 @@
     "polaris.shopify.com"
   ],
   "scripts": {
-    "build": "yarn workspaces run build",
+    "build": "turbo run build",
+    "dev": "turbo run dev --no-cache --parallel --continue",
```
Comment (aaronccasanova) [inline package.json]:
```suggestion
    "dev": "turbo run dev --parallel",
```

You already have `cache: false` set it the `turbo.json` config for the `dev` pipeline.

https://turborepo.org/docs/reference/command-line-reference#--continue
> By default, specifying the `--parallel` flag will automatically set `--continue` to `true`

File: polaris-for-vscode/package.json [hunk 2]
```diff
@@ -32,10 +32,12 @@
   "scripts": {
     "build": "tsc -b",
     "dev": "tsc -b -w",
+    "lint": "TIMING=1 eslint server client",
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
```
Comment (aaronccasanova) [inline polaris-for-vscode/package.json]:
```suggestion
    "clean": "rm -rf .turbo node_modules dist",
```

File: polaris-icons/package.json [hunk 3]
```diff
@@ -7,8 +7,10 @@
   "module": "dist/index.mjs",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "yarn run rollup -c",
-    "test": "jest"
+    "build": "rollup -c",
+    "dev": "rollup -w",
```
Comment (aaronccasanova) [inline polaris-icons/package.json]:
```suggestion
    "dev": "rollup -c -w",
```

File: polaris-icons/package.json [hunk 4]
```diff
@@ -7,8 +7,10 @@
...
+    "test": "jest",
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
```
Comment (aaronccasanova) [inline polaris-icons/package.json]:
```suggestion
    "clean": "rm -rf .turbo node_modules dist"
```

File: polaris-react/package.json [hunk 5]
```diff
@@ -52,12 +52,12 @@
...
     "storybook": "yarn gen-media-queries && start-storybook -p 6006 --quiet",
     "storybook:build": "yarn gen-media-queries && build-storybook -o build-internal/storybook/static",
-    "test": "loom test",
+    "test": "loom test --no-watch",
     "validate-tokens": "ajv validate -s src/tokens/schemas/token-group.json -d 'src/tokens/token-groups/*.json'",
     "version": "node ./scripts/readme-update-version"
   },
   "dependencies": {
-    "@shopify/polaris-icons": "^4.18.2",
+    "@shopify/polaris-icons": "*",
```
Comment (aaronccasanova) [inline polaris-react/package.json:60]: **question:** Is `polaris-icons` fully transitioned over to the `polaris` monorepo?

File: polaris-tokens/package.json [hunk 6]
```diff
@@ -18,8 +18,11 @@
     "build": "run-s build:*",
     "build:js": "rollup -c",
     "build:assets": "ts-node --skip-project scripts/index.ts",
-    "build:types": "tsc",
-    "test": "jest"
+    "build:types": "tsc --emitDeclarationOnly",
+    "dev": "rollup -w",
```
Comment (aaronccasanova) [inline polaris-tokens/package.json]:
```suggestion
    "dev": "rollup -c -w",
```

File: polaris-tokens/package.json [hunk 7]
```diff
@@ -18,8 +18,11 @@
...
+    "lint": "TIMING=1 eslint src",
+    "test": "jest",
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
```
Comment (aaronccasanova) [inline polaris-tokens/package.json]:
```suggestion
    "clean": "rm -rf .turbo node_modules dist"
```

File: polaris.shopify.com/package.json [hunk 8]
```diff
@@ -5,15 +5,17 @@
   "scripts": {
     "build": "next build",
     "dev": "next dev",
-    "start": "next start",
+    "start": "next start ",
```
Comment (aaronccasanova) [inline polaris.shopify.com/package.json:8]:
```suggestion
    "start": "next start",
```

File: polaris.shopify.com/package.json [hunk 9]
```diff
@@ -5,15 +5,17 @@
...
+    "lint": "TIMING=1 eslint src",
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf .next",
```
Comment (aaronccasanova) [inline polaris.shopify.com/package.json]:
```suggestion
    "clean": "rm -rf .turbo node_modules .next",
```

File: stylelint-polaris/package.json [hunk 10]
```diff
@@ -24,11 +24,12 @@
...
   ],
-  "main": "index.js",
+  "main": "src/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "echo no build script",
-    "test": "node tests/index.js"
+    "lint": "TIMING=1 eslint src",
+    "test": "node tests/index.js",
+    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
```
Comment (aaronccasanova) [inline stylelint-polaris/package.json]:
```suggestion
    "clean": "rm -rf .turbo node_modules dist"
```

**Review by aaronccasanova — COMMENTED** [review]
So excited for this addition! Left some initial feedback to consider and plan to pull down the branch right now to tophat 🎩

**Review by aaronccasanova — APPROVED** [review]
Tophatted this locally and the caching applied to subsequent builds is incredible!

File: polaris-react/package.json (hunk 5 above)
Comment (sam-b-rose) [reply polaris-react/package.json:60]: Good question 😅 we need it to be bc with the way yarn workspaces works, it keeps resolving to the local version. Didn't seem to break vis diff in Chromatic I don't think

**Comment by sam-b-rose** [comment]
Thanks for the review @aaronccasanova! Addressed all feedback, great clean up suggestions ✨

**Pull request merged**